### DISCUSSION TITLE
Is there a way to make it so that mpvnet automatically flashes chapter name when chapter of a video changes?

### DISCUSSION THREAD

**Discussion by uugaabuugaa** [mpvnet-player/mpv.net#695] [Q&A] [answered]
I found this on reddit:
```
local function showChapterTitle()
    local chapterTitle = mp.get_property_osd("chapter-metadata/by-key/title")
    mp.osd_message(chapterTitle, 3)
end

mp.observe_property("chapter", nil, showChapterTitle)
```

But the problem is if a video doesn't have any chapter it will show the file's title again. So I get two file-title flashes at the start when there's no chapter in the video. Any way to make it so the chapter name is shown only when there are multiple chapters present in the video?

**Comment by Sneakpeakcss** [comment]
> But the problem is if a video doesn't have any chapter it will show the file's title again. So I get two file-title flashes at the start when there's no chapter in the video.

This shouldn't be happening. Even with this script it only displays chapter name when it actually encounters one, without any additional title flashes on files with/without them. Sounds like there's some configuration error on your end. 

You can try adding `osd-playing-msg=""` in `mpv.conf` and see what happens when the default fileload msg is removed, or even check if you don't have some profiles that already use it and clash with each other.

**Reply by uugaabuugaa** [reply]
>  Sounds like there's some configuration error on your end.

Tried installing a fresh portable instance of mpvnet with only this script but the issue was still occurring. However, `osd-playing-msg=""` has almost solved the issue and chapter-title osd only shows up when encountering one, even in my original mpvnet.

It "almost solved the issue" because yesterday I was going through some of my other scripts to see if something related was there and found out putting  `mp.osd_message('Chapter: '.. mp.command_native({'expand-text', '${chapter}'}), 3)` instead of `mp.osd_message(chapterTitle, 3)`, shows the chapters just the way I want it to like this: `Chapter: (1) Intro` instead of only showing `Intro`. But the problem is when I'm using this command, even with  `osd-playing-msg=""` in my `mpv.conf` it still shows `Chapter: (Unavailable)` at startup of mpvnet and start of playback of files without a chapter. I'll go with `mp.osd_message(chapterTitle, 3)` for now since showing chapters when I want to is more important than respecting the format. But I think there has to be way to make an argument so that if chapter is unavailable it shows nothing at startup and loading of file without chapters although I have no idea how.

**Reply by Sneakpeakcss** [reply, marked as answer]
> Tried installing a fresh portable instance of mpvnet with only this script but the issue was still occurring. 

I assume that you made sure that nothing loaded, since even portable version loads cfg/scripts from `mpv.net` directory if it's present in `appdata/roaming`. The script you posted has no right to interfere with files that have no chapters, so any flashing in such files is something beyond weird. The only title flash this could create is only if there was an intro chapter at the beginning that displayed for split of a second before `osd-playing-msg` overwrites it with the filename.

> (…) But I think there has to be way to make an argument so that if chapter is unavailable it shows nothing at startup and loading of file without chapters although I have no idea how.

Avoiding displaying the `(unavailable)` message is as simple as this:

```lua
local function showChapterTitle()
    local chapterExists = mp.get_property("chapter")
    local idle = mp.get_property_native("idle-active")
    
    if chapterExists and not idle then
        mp.osd_message('Chapter: '.. mp.command_native({'expand-text', '${chapter}'}), 3)
    end
end
mp.observe_property("chapter", nil, showChapterTitle)
```

However, this will still have the issue with the mentioned files containing intro chapters that start at `00:00:00`, it will display it for a split of a second and then default `osd-playing-msg` will overwrite it because osd filename message is a bit delayed. Any files without chapters have to behave as mentioned above. 

You can use `print-text ${chapters}` in the console to make sure that those files definitely do not have any chapters. If they show `0` and those two title flashes are still happening, then you might need to post a video of it because this doesn’t happen on my end with any of those scripts.

**Reply by uugaabuugaa** [reply]
> ```
> 
> local function showChapterTitle()
>     local chapterExists = mp.get_property("chapter")
>     local idle = mp.get_property_native("idle-active")
>     
>     if chapterExists and not idle then
>         mp.osd_message('Chapter: '.. mp.command_native({'expand-text', '${chapter}'}), 3)
>     end
> end
> mp.observe_property("chapter", nil, showChapterTitle)
> 
> ```

This is exactly what I needed. Thank you so much!! Tried with files that have chapters, doesn't have any chapters and so far it's working exactly as I require it to. Bless you my guy.

> it will display it for a split of a second and then default `osd-playing-msg` will overwrite it because osd filename message is a bit delayed

Haven't had this issue yet maybe because I kept the `osd-playing-msg=""` in `mpv.conf` and the only title flash I get is from uosc's top-bar flash that happens on every file loading instances so I'm good so far. For clarification, I used to have 3 title flash on files without chapters with the previous script; one for uosc's top-bar flash, two from native mpvnet. So now I'm happier with only seeing uosc's top-bar flash and chapter title osd depending on if it's available or not.

> I assume that you made sure that nothing loaded, since even portable version loads cfg/scripts from `mpv.net` directory if it's present in `appdata/roaming`

Yes nothing else loaded that time because I use the portable version as default already so my `appdata/roaming/mpv.net` doesn't have any scripts there.
